From: Peter Zijlstra <a.p.zijlstra@chello.nl>
Subject: perf, arch: Cleanup perf-pmu init vs lockup-detector
References: fate#311939, bnc#704917
Patch-mainline: v2.6.38-rc1
Signed-off-by: Tony Jones <tonyj@suse.de>


Signed-off-by: Youquan Song <youquan.song@intel.com>

The perf hardware pmu got initialized at various points in the boot,
some before early_initcall() some after (notably arch_initcall).

The problem is that the NMI lockup detector is ran from early_initcall()
and expects the hardware pmu to be present.

Sanitize this by moving all architecture hardware pmu implementations to
initialize at early_initcall() and move the lockup detector to an explicit
initcall right after that.

Youquan: we need add this piece code because when init_hw_perf_events() run
in function identify_boot_cpu(), some data structure about CPU topology of CPU
 core and thread is not initialized, so if without the piece of code, the 
kernel with path "Add support for supplementary event registers" a7e3ed1e4 has
 issue to hange at boot. We only backport x86 desirable code which does not make
 any change to other architecutre, so they are not affected. 

Cc: paulus <paulus@samba.org>
Cc: davem <davem@davemloft.net>
Cc: Michael Cree <mcree@orcon.net.nz>
Cc: Deng-Cheng Zhu <dengcheng.zhu@gmail.com>
Acked-by: Paul Mundt <lethal@linux-sh.org>
Acked-by: Will Deacon <will.deacon@arm.com>
Signed-off-by: Peter Zijlstra <a.p.zijlstra@chello.nl>
LKML-Reference: <1290707759.2145.119.camel@laptop>
Signed-off-by: Ingo Molnar <mingo@elte.hu>

Automatically created from "patches.suse/intel-perf-0108-plus-pmu.patch" by xen-port-patches.py

--- sle11sp2-2011-07-15.orig/arch/x86/kernel/cpu/common-xen.c	2011-07-15 14:16:03.000000000 +0200
+++ sle11sp2-2011-07-15/arch/x86/kernel/cpu/common-xen.c	2011-07-15 14:21:56.000000000 +0200
@@ -945,7 +945,6 @@ void __init identify_boot_cpu(void)
 #else
 	vgetcpu_set_mode();
 #endif
-	init_hw_perf_events();
 }
 
 #ifdef CONFIG_XEN
